File: benchmarks/pandas/bench_concat_options.py
"""
Benchmark: pandas concat with join="inner" and ignore_index=True options.
Outputs JSON: {"function": "concat_options", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd

ROWS = 50_000
WARMUP = 5
ITERATIONS = 20

df1 = pd.DataFrame({
    "a": [i * 1.0 for i in range(ROWS)],
    "b": [i * 2.0 for i in range(ROWS)],
    "c": [i * 3.0 for i in range(ROWS)],
})
df2 = pd.DataFrame({
    "a": [i * 1.5 for i in range(ROWS)],
    "b": [i * 2.5 for i in range(ROWS)],
    "d": [i * 4.0 for i in range(ROWS)],
})

for _ in range(WARMUP):
    pd.concat([df1, df2], join="inner", ignore_index=True)
    pd.concat([df1, df2], join="outer", ignore_index=True)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    pd.concat([df1, df2], join="inner", ignore_index=True)
    pd.concat([df1, df2], join="outer", ignore_index=True)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
mean_ms = total_ms / ITERATIONS
print(json.dumps({"function": "concat_options", "mean_ms": round(mean_ms, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))

File: benchmarks/tsb/bench_concat_options.ts
/**
 * Benchmark: concat with join="inner" and ignoreIndex=true options.
 * Outputs JSON: {"function": "concat_options", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame, concat } from "../../src/index.ts";

const ROWS = 50_000;
const WARMUP = 5;
const ITERATIONS = 20;

// Two DataFrames with partial column overlap (inner join drops non-shared columns)
const df1 = DataFrame.fromColumns({
  a: Array.from({ length: ROWS }, (_, i) => i * 1.0),
  b: Array.from({ length: ROWS }, (_, i) => i * 2.0),
  c: Array.from({ length: ROWS }, (_, i) => i * 3.0),
});
const df2 = DataFrame.fromColumns({
  a: Array.from({ length: ROWS }, (_, i) => i * 1.5),
  b: Array.from({ length: ROWS }, (_, i) => i * 2.5),
  d: Array.from({ length: ROWS }, (_, i) => i * 4.0),
});

for (let i = 0; i < WARMUP; i++) {
  concat([df1, df2], { join: "inner", ignoreIndex: true });
  concat([df1, df2], { join: "outer", ignoreIndex: true });
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  concat([df1, df2], { join: "inner", ignoreIndex: true });
  concat([df1, df2], { join: "outer", ignoreIndex: true });
  times.push(performance.now() - t0);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "concat_options",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);
